src/cargo/core/compiler/custom_build.rs

```diff
@@ -201,6 +202,31 @@ fn build_work(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Job> {
         .env("HOST", &bcx.host_triple())
         .env("RUSTC", &bcx.rustc().path)
         .env("RUSTDOC", &*bcx.config.rustdoc()?)
+        .env(
+            "CARGO_MODE",
+            match unit.root_mode {
+                CompileMode::Test => "Test",
+                CompileMode::Build => "Build",
+                CompileMode::Check { test } => {
+                    if test {
+                        "Check_test"
+                    } else {
+                        "Check"
+                    }
+                }
+                CompileMode::Bench => "Bench",
+                CompileMode::Doc { deps } => {
+                    if deps {
```

Comment:
This shouldn't matter to the build script, no dependencies will be built *after* it runs.